server.js

```diff
@@ -731,166 +731,127 @@ app.post('/submit-match',  asyncMiddleware( async (req, res, next) => {
-                                // at the end and not try to reduce database lookups?
-                                //
-                                // Ok new problem, during the async stuff later more requests come in and so network2=null matches
-                                // don't face right opponent. Lets do a sync copy if this was in fact a new best network
-                                // situation for the current active match in queue.
-                                //
-                                // added !is_test flag
-                                //
-                                if (req.body.loserhash == best_network_hash && !pending_matches[pending_matches.length - 1].is_test) {
-                                    new_best_network_flag = true;
-
-                                    fs.copyFileSync(__dirname + '/network/' + req.body.winnerhash + '.gz', __dirname + '/network/best-network.gz');
-                                    console.log("New best network copied from (fast check): " + __dirname + '/network/' + req.body.winnerhash + '.gz');
-                                }
-                            }
-                        } else {
-                            // network1 was the winner but it was no longer at the end of the pending_match queue.
-                            //
-                        }
+            var sgfhash = checksum(sgfbuffer, 'sha256');
+
+            // upload match game to database
+            var dbres = await db.collection("match_games").updateOne(
+                { sgfhash: sgfhash },
+                {
+                    $set: {
+                        ip: req.ip, winnerhash: req.body.winnerhash, loserhash: req.body.loserhash, sgf: sgfbuffer.toString(),
+                        options_hash: req.body.options_hash,
+                        clientversion: Number(req.body.clientversion), winnercolor: req.body.winnercolor,
+                        movescount: (req.body.movescount ? Number(req.body.movescount) : null),
+                        score: req.body.score,
+                        random_seed: req.body.random_seed
                     }
-                }
+                },
+                { upsert: true }
+            );
+
+            // Not inserted, we got duplicate sgfhash, abort!
+            if (!dbres.upsertedId) {
+                return reject('Upload match with duplicate sgf');
             }
-        );
-      }
+
+            console.log(req.ip + " (" + req.headers['x-real-ip'] + ") " + " uploaded match " + sgfhash);
+            res.send("Match data " + sgfhash + " stored in database\n");
+            resolve();
+        });
     });
 
+    // update match result 
+    if (match.network1 == req.body.winnerhash)
+        match.network1_wins++;
+    else
+        match.network1_losses++;
+
+    // save to database
+    await db.collection('matches').updateOne(
+        { _id: match._id },
+        { $set: match }
```

Comment:
Seems like concurrent match submissions could end up reading out a match from the db and after awaiting for sgfhash, both try to update to a locally computed `network1_wins/losses`. Is there a way to have the update tell the db to increment a field instead of passing it a value to set?